includes/class-wc-geolocation.php

```diff
@@ -255,40 +255,35 @@ public static function update_database() {
 
 		require_once ABSPATH . 'wp-admin/includes/file.php';
 
-		$upload_dir        = wp_upload_dir();
-		$tmp_database_path = download_url( self::GEOLITE2_DB );
+		$database             = 'GeoLite2-Country.mmdb';
+		$target_database_path = self::get_local_database_path();
+		$tmp_database_path    = download_url( self::GEOLITE2_DB );
 
 		if ( ! is_wp_error( $tmp_database_path ) ) {
 			try {
-				// GeoLite2 database name.
-				$database  = 'GeoLite2-Country.mmdb';
-				$dest_path = trailingslashit( $upload_dir['basedir'] ) . $database;
+				WP_Filesystem();
+
+				global $wp_filesystem;
+
+				// Make sure target dir exists.
+				$wp_filesystem->mkdir( dirname( $target_database_path ) );
 
 				// Extract files with PharData. Tool built into PHP since 5.3.
 				$file      = new PharData( $tmp_database_path ); // phpcs:ignore PHPCompatibility.Classes.NewClasses.phardataFound
 				$file_path = trailingslashit( $file->current()->getFileName() ) . $database;
+				$file->extractTo( dirname( $tmp_database_path ), $file_path, true );
 
-				// Extract under uploads directory.
-				$file->extractTo( $upload_dir['basedir'], $file_path, true );
-
-				// Remove old database.
-				@unlink( $dest_path ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.VIP.FileSystemWritesDisallow.file_ops_unlink
-
-				// Copy database and delete tmp directories.
-				@rename( trailingslashit( $upload_dir['basedir'] ) . $file_path, $dest_path ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.VIP.FileSystemWritesDisallow.file_ops_rename
-				@rmdir( trailingslashit( $upload_dir['basedir'] ) . $file->current()->getFileName() ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.VIP.FileSystemWritesDisallow.directory_rmdir
-
-				// Set correct file permission.
-				@chmod( $dest_path, 0644 ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.VIP.FileSystemWritesDisallow.chmod_chmod
+				// Move file and delete temp.
+				$wp_filesystem->move( trailingslashit( dirname( $tmp_database_path ) ) . $file_path, $target_database_path, true );
+				$wp_filesystem->unlink( trailingslashit( dirname( $tmp_database_path ) ) . $file->current()->getFileName() );
+				$wp_filesystem->unlink( $tmp_database_path );
 			} catch ( Exception $e ) {
 				$logger->notice( $e->getMessage(), array( 'source' => 'geolocation' ) );
 
 				// Reschedule download of DB.
 				wp_clear_scheduled_hook( 'woocommerce_geoip_updater' );
 				wp_schedule_event( strtotime( 'first tuesday of next month' ), 'monthly', 'woocommerce_geoip_updater' );
 			}
-
-			@unlink( $tmp_database_path ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.VIP.FileSystemWritesDisallow.file_ops_unlink
```

Comment:
If there is any trouble extracting or something else, it will keep the file in `/tmp`. That's why I kept this `unlink` here.